File: 905-sort-array-by-parity/Case2.java
//Time: 1 ms (96.61%), Space: 48.7 MB (15.61%) - LeetHub

class Solution {
    public int[] sortArrayByParity(int[] nums) {
        int[] ans = new int[nums.length];
        int lo = 0, hi = nums.length - 1;
        
        for (int i = 0; i < nums.length; i++) 
            if ((nums[i] & 1) == 1) 
                ans[hi--] = nums[i];
            else
                ans[lo++] = nums[i];
        
        return ans;
    }
}

File: 47-permutations-ii/Case2.cpp
//Time: 0 ms (100.00%), Space: 8.4 MB (95.68%) - LeetHub

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        do {
            ans.push_back(nums);
            nextPermutation(nums);
        } while (nums != ans[0]);
        
        return ans;
    }
    
    void nextPermutation(vector<int>& nums) {
        if (nums.size() < 2) 
            return;
        
        int i = nums.size() - 1, j = nums.size();
        while (--i >= 0 && nums[i] >= nums[i + 1]);
        
        if (i >= 0) {
            while (nums[i] >= nums[--j]);
            swap(nums[i], nums[j]);
        }
        reverse(nums.begin() + i + 1, nums.end());
    }
};
